homeassistant/util/timeout.py

```diff
@@ -0,0 +1,458 @@
+            return
+        self._state = _State.FREEZE
+        self._stop_timer()
+
+    def reset(self) -> None:
+        """Reset timer after freeze."""
+        if self._state != _State.FREEZE:
+            return
+        self._state = _State.ACTIVE
+        self._start_timer()
+
+
+class ZoneTimeout:
+    """Async zone based timeout handler."""
+
+    def __init__(self):
+        """Initalize ZoneTimeout handler."""
+        self._loop: asyncio.AbstractEventLoop = asyncio.get_running_loop()
+        self._zones: Dict[str, _Zone] = {}
+        self._globals: List[_TaskGlobal] = []
+        self._freezes: List[_FreezeGlobal] = []
+
+    @property
+    def zones_done(self) -> bool:
+        """Return True if all zones are finish."""
+        return not bool(self._zones)
+
+    @property
+    def freezes_done(self) -> bool:
+        """Return True if all freezes are finish."""
+        return not self._freezes
+
+    @property
+    def zones(self) -> Dict[str, _Zone]:
+        """Return all Zones."""
+        return self._zones
+
+    @property
+    def global_tasks(self) -> List[_TaskGlobal]:
+        """Return all global Tasks."""
+        return self._globals
+
+    @property
+    def global_freezes(self) -> List[_FreezeGlobal]:
+        """Return all global Freezes."""
+        return self._freezes
+
+    def drop_zone(self, zone_name: str) -> None:
+        """Drop a zone out of scope."""
+        self._zones.pop(zone_name, None)
+        if self._zones:
+            return
+
+        # Signal Global task, all zones are done
+        for task in self._globals:
+            task.zones_done_signal()
+
+    def async_timeout(
+        self, timeout: float, zone_name: str = ZONE_GLOBAL, cool_down: float = 0
+    ) -> Union[_TaskZone, _TaskGlobal]:
+        """Timeout based on a zone.
+
+        For using as Async Context Manager.
+        """
+        current_task: asyncio.Task[Any] = asyncio.current_task()
+
+        # Global Zone
+        if zone_name == ZONE_GLOBAL:
+            task = _TaskGlobal(self, current_task, timeout, cool_down)
+            return task
+
+        # Zone Handling
+        if zone_name in self.zones:
+            zone: _Zone = self.zones[zone_name]
+        else:
+            self.zones[zone_name] = zone = _Zone(self, zone_name, timeout)
+
+        # Create Task
+        task = _TaskZone(zone, current_task)
+        return task
+
+    def freeze(self, zone_name: str = ZONE_GLOBAL) -> Union[_FreezeZone, _FreezeGlobal]:
+        """Freeze all timer until job is done.
+
+        For using as (Async) Context Manager.
+        """
+        # Global Freeze
+        if zone_name == ZONE_GLOBAL:
+            return _FreezeGlobal(self, self._loop)
+
+        # Zone Freeze
+        if zone_name in self.zones:
+            zone: _Zone = self.zones[zone_name]
+            return _FreezeZone(zone, self._loop)
+
+        # Should not happens...
+        return _FreezeGlobal(self, self._loop)
```

Comment:
```suggestion
        # Should only happen in tests.
        _LOGGER.debug("Cannot freeze zone: %s because no timeout is set.")
        return _FreezeNull()
```